docs: add Clickhouse documentation (#1757)
emqtt/emq-docs-cn: en_US/data-integration/data-bridge-clickhouse.md

# Clickhouse 

EMQX supports integration with [Clickhouse](https://clickhouse.com/) so you can forward messages and events to Clickhouse. 

{% emqxce %}
:::tip
The Clickhouse bridge is an EMQX Enterprise Edition feature. EMQX Enterprise Edition provides comprehensive coverage of key business scenarios, rich data integration, product-level reliability, and 24/7 global technical support. Experience the benefits of this [enterprise-ready MQTT messaging platform](https://www.emqx.com/en/try?product=enterprise) today.
:::
{% endemqxce %}

::: tip Prerequisites

- Knowledge about EMQX data integration [rules](./rules.md)

- Knowledge about [data bridges](./data-bridges.md)

> - Basic knowledge about running commands in a UNIX terminal

Reviewer: can we add a link here so they can know where to learn these knowledge?

Author: I think we can remove the word UNIX to make it sound less scary but I don't think a link will add much value. System administrators reading this document will know what a terminal is. So I think we can remove this line.

# ClickHouse 

EMQX supports integration with [ClickHouse](https://clickhouse.com/) so you can forward messages and events to ClickHouse. 

{% emqxce %}
:::tip
The ClickHouse bridge is an EMQX Enterprise Edition feature. EMQX Enterprise Edition provides comprehensive coverage of key business scenarios, rich data integration, product-level reliability, and 24/7 global technical support. Experience the benefits of this [enterprise-ready MQTT messaging platform](https://www.emqx.com/en/try?product=enterprise) today.
:::
{% endemqxce %}

::: tip Prerequisites

- Knowledge about EMQX data integration [rules](./rules.md)

- Knowledge about [data bridges](./data-bridges.md)

- Basic knowledge of UNIX terminal and commands 

:::

## Feature List

- [Connection pool](./data-bridges.md)
- [Async mode](./data-bridges.md)
- [Batch mode](./data-bridges.md)
- [Buffer mode](./data-bridges.md)

## Quick Start Tutorial

> This section introduces how to use the ClickHouse bridge with a practical tutorial, covering topics like how to create a Clickouse server, how to set up a bridge, and how to set up a rule for forwarding data to the bridge and testing that it all works. 

Author: Clickouse server -> ClickHouse server

> This section introduces how to use the ClickHouse bridge with a practical tutorial, covering topics like how to create a Clickouse server, how to set up a bridge, and how to set up a rule for forwarding data to the bridge and testing that it all works. 

Reviewer: ```suggestion
This section introduces how to use the ClickHouse bridge with a practical tutorial, covering topics like how to create a ClickHouse server, how to set up a bridge, and how to set up a rule for forwarding data to the bridge and testing that it all works. 
```

   docker run \
   --rm \
   -e CLICKHOUSE_DB=mqtt_data \
   -e CLICKHOUSE_USER=emqx \
   -e CLICKHOUSE_DEFAULT_ACCESS_MANAGEMENT=1 \
   -e CLICKHOUSE_PASSWORD=public \
   -p 18123:8123 \
   -p19000:9000 \
   --ulimit nofile=262144:262144 \
   -v ./init.sql:/docker-entrypoint-initdb.d/init.sql \
   clickhouse/clickhouse-server
   ```

You can find more information about running ClickHouse in docker [on dockerhub](https://hub.docker.com/r/clickhouse/clickhouse-server).

### Create a ClickHouse Data Bridge

Then you can start to create an EMQX data bridge to ClickHouse.

1. Go to EMQX Dashboard, click **Data Integration** -> **Data Bridge**.

2. Click **Create** on the top right corner of the page.

3. In the **Create Data Bridge** page, click to select **ClickHouse**, and then click **Next**.

4. Input a name for the data bridge. The name should be a combination of upper/lower case letters and numbers.

5. Input the connection information:
   *  **Server URL**: Input **http://127.0.0.1:18123**, or the actual URL if the ClickHouse server is running remotely.
   * **Database Name**: Input **mqtt_data**.
   * **Username**: Input **emqx**.
   * **Password**: Input **public**
   

> 7. **Batch Separator** (optional): for separate multiple items, default:  `,`  . You can customize the separate based on the data input/out format you defined in ClickHouse, for example, if [JSONCompactEachRow` format](https://clickhouse.com/docs/en/interfaces/formats#jsoncompacteachrow)  is used, should leave it blank. 

Author: I don't think this is easy to understand for anyone that does not already know what it is. How would the user now at this point what it means to separate multiple items. How would the user even know what items refer to? I think some context has been lost from the original description (for example, that this is something that is only relevant to batching).

> You can customize the separate based on the data input/out format you defined in ClickHouse

I don't think this is a good way of describing it. The format is not specified inside clickhouse. It is specified in the SQL statement. I think the original text had a reference to the clickhouse syntax for INSERT statements. 

the separate -> the separator

My suggestion is to put this description as a separate section as it originally was and try to make it easier to understand. Perhaps it would be useful with a an example showing an SQL statement that use a different separator.

Perhaps you can keep this item but rewrite it to something like this:

In this example you can keep the default value "`,`" for this setting. This setting only needs to be changed if you enable [batching](link to batching desc) for the bridge and if you specify an alternative format with [ClickHouse's FORMAT syntax](link to clickhouse's doc for the INSERT statement). For further info see [advanced settings](link to advanced settings).

> 7. **Batch Separator** (optional): for separate multiple items, default:  `,`  . You can customize the separate based on the data input/out format you defined in ClickHouse, for example, if [JSONCompactEachRow` format](https://clickhouse.com/docs/en/interfaces/formats#jsoncompacteachrow)  is used, should leave it blank. 

Reviewer: ```suggestion
7. **Batch Separator** (optional): In this example, you can keep the default value ",". This setting only needs to be changed if you enable [batching](./data-bridges.md) for the bridge and if you specify an alternative format with [ClickHouse's FORMAT syntax](https://clickhouse.com/docs/en/sql-reference/statements/insert-into).
```

> 7. Insert the following text in the **SQL Template** text area (You can use [Rule Engine](./rules.md)  to ensure that the specified SQL statements are escaped and not vulnerable to SQL injection attacks.):

Author: > You can use [Rule Engine](./rules.md) to ensure that the specified SQL statements are escaped and not vulnerable to SQL injection attacks.

->

You can use [Rule Engine](./rules.md) to ensure that strings in the specified SQL statement are escaped so the SQL statement is not vulnerable to SQL injection attacks.

> 7. Insert the following text in the **SQL Template** text area (You can use [Rule Engine](./rules.md)  to ensure that the specified SQL statements are escaped and not vulnerable to SQL injection attacks.):

Reviewer: ```suggestion
7. Insert the following text in the **SQL Template** text area (You can use [Rule Engine](./rules.md) to ensure that strings in the specified SQL statement are escaped so the SQL statement is not vulnerable to SQL injection attacks.):
```

   ```
   The `${payload.data}` and `${payload.timestamp}` are placeholders for the data and timestamp of the message coming from the [rule](#create-a-rule-for-our-clickhouse-bridge) you will configure later. The placeholders will be replaced by the actual data before the message is sent to the ClickHouse server.
   
7. Advanced settings (optional):  Choose whether to use sync or async query mode etc, or batch mode. For details about these advanced features, see the [general data bridge documentation](./data-bridges.md).

8. Before clicking **Create**, you can click **Test Connection** to test that the bridge can connect to the ClickHouse server.

10. Then click **Create** to finish the creation of the data bridge.

Now the ClickHouse data bridge should appear in the data bridge list (**Data Integration** -> **Data Bridge**) with **Resource Status** as **Connected**. You can continue to create a rule to forward data to the new ClickHouse bridge. 

### Create a Rule for the ClickHouse Bridge

1. Go to EMQX Dashboard, click **Data Integration** -> **Rules**.

2. Click **Create** on the top right corner of the page.

3. Input, for example, `my_rule` as the rule ID.

> 5. Input the following SQL statements in the SQL editor, which will forward the MQTT messages matching the topic pattern `t/#` to the newly created ClickHouse bridge. 

Author: this is one SQL statement

SQL statements -> statement

you can skip SQL as well (it does not add any value)

Author: "to the newly created ClickHouse bridge" I think this part can be deleted.

> 5. Input the following SQL statements in the SQL editor, which will forward the MQTT messages matching the topic pattern `t/#` to the newly created ClickHouse bridge. 

Reviewer: ```suggestion
5. Input the following statement in the SQL editor, which will forward the MQTT messages matching the topic pattern `t/#`. 
```

   
   ```sql
   SELECT 
     payload as data,
     now_timestamp() as timestamp
   FROM
     "t/#"
   ```
   
5. Then click the **Add Action** button, select **Forwarding with Data Bridge** from the dropdown list and then select the data bridge you just created under **Data bridge**.  

> 7. Click the **Add** button to finish the setup. Click the **Create** button at the page bottom to finish the setup. 

Author: I suggest that this is broken into:

```
7. Click the **Add** button to finish the setup of the action
8. . Click the **Create** button at the page bottom to save the rule.
```

> 7. Click the **Add** button to finish the setup. Click the **Create** button at the page bottom to finish the setup. 

Reviewer: ```suggestion
7. Click the **Add** button to finish the setup. 
8. Click the **Create** button at the page bottom to finish the setup. 
```

Now a rule to forward data to ClickHouse via a ClickHouse bridge is created. You can click **Data Integration** -> **Flows** to view the topology. It can be seen that the messages under topic `t/#`  are sent and saved to ClickHouse. 

> ### Test Rule and Bridge

Author: Test the Rule and the Bridge

?

> ### Test Rule and Bridge

Reviewer: ```suggestion
### Test the Rule and Bridge
```

# ClickHouse 

> EMQX supports integration with [ClickHouse](https://clickhouse.com/) so you can forward messages and events to ClickHouse. 

Reviewer 2: It is more special and can be introduced more
```suggestion
ClickHouse is a high-performance distributed database management system designed for processing large-scale data. It features excellent query performance, a flexible data model, and scalable distributed architecture, making it suitable for various data analytics scenarios.

EMQX supports integration with [ClickHouse](https://clickhouse.com/) so you can save messages and events data to ClickHouse. 
```

- Knowledge about [data bridges](./data-bridges.md)

- Basic knowledge of UNIX terminal and commands 

:::

## Feature List

- [Connection pool](./data-bridges.md)
- [Async mode](./data-bridges.md)
- [Batch mode](./data-bridges.md)
- [Buffer mode](./data-bridges.md)

## Quick Start Tutorial

This section introduces how to use the ClickHouse bridge with a practical tutorial, covering topics like how to create a ClickHouse server, how to set up a bridge, and how to set up a rule for forwarding data to the bridge and testing that it all works. 

This tutorial assumes that you run both EMQX and ClickHouse on the local machine. If you have ClickHouse and EMQX running remotely, please adjust the settings accordingly.

### Start a ClickHouse Server

This section introduces how to start a ClickHouse server using [Docker](https://www.docker.com/). 

1. Create a file called `init.sql` with the following initialization SQL statements, which will help to initialize the database when the container starts up.

   ```bash
   cat >init.sql <<SQL_INIT
   CREATE DATABASE IF NOT EXISTS mqtt_data;
   CREATE TABLE IF NOT EXISTS mqtt_data.messages (
       data String,

>        arrived BIGINT

Reviewer 2: More strict data types
```suggestion
       arrived UnixTimestamp
```

>    ) ENGINE = Memory;

Reviewer 2: ```suggestion
   ) ENGINE = MergeTree();
```

   SQL_INIT
   ```

2. Then, start a ClickHouse server using the following command, which defines the database name, port number, user name and password. And it will also mount the `init.sql` file in the current directory to the docker directory.

   ```bash
   docker run \
   --rm \
   -e CLICKHOUSE_DB=mqtt_data \
   -e CLICKHOUSE_USER=emqx \
   -e CLICKHOUSE_DEFAULT_ACCESS_MANAGEMENT=1 \
   -e CLICKHOUSE_PASSWORD=public \
   -p 18123:8123 \

>    -p19000:9000 \

Reviewer 2: ```suggestion
   -p 19000:9000 \
```

3. In the **Create Data Bridge** page, click to select **ClickHouse**, and then click **Next**.

4. Input a name for the data bridge. The name should be a combination of upper/lower case letters and numbers.

5. Input the connection information:
   *  **Server URL**: Input **http://127.0.0.1:18123**, or the actual URL if the ClickHouse server is running remotely.
   * **Database Name**: Input **mqtt_data**.
   * **Username**: Input **emqx**.
   * **Password**: Input **public**
   
7. **Batch Separator** (optional): In this example, you can keep the default value ",". This setting only needs to be changed if you enable [batching](./data-bridges.md) for the bridge and if you specify an alternative format with [ClickHouse's FORMAT syntax](https://clickhouse.com/docs/en/sql-reference/statements/insert-into).
   
7. Insert the following text in the **SQL Template** text area (You can use [Rule Engine](./rules.md) to ensure that strings in the specified SQL statement are escaped so the SQL statement is not vulnerable to SQL injection attacks.):
   
   ```sql
   INSERT INTO messages(data, arrived) VALUES ('${data}', ${timestamp})
   ```

>    The `${payload.data}` and `${payload.timestamp}` are placeholders for the data and timestamp of the message coming from the [rule](#create-a-rule-for-our-clickhouse-bridge) you will configure later. The placeholders will be replaced by the actual data before the message is sent to the ClickHouse server.

Reviewer 2: ```suggestion
   The `${data}` and `${timestamp}` are placeholders for the data and timestamp of the message coming from the [rule](#create-a-rule-for-our-clickhouse-bridge) you will configure later. The placeholders will be replaced by the actual data before the message is sent to the ClickHouse server.
```
